### ISSUE TITLE
NoMethodError for update

### ISSUE THREAD

**Issue by novakw267** [ga-wdi-boston/full-stack-project#706]
I am getting a 
```
NoMethodError (undefined method `update' for nil:NilClass):
```
 when I try to update my profile on the front end, I have tested it on the back end in node and was successful so that leads me to say it's probably a front end issue. 

api for update
```
const updateProfile = function (id, data) {
  return $.ajax({
    url: config.apiOrigin + "/profiles/" + id,
    method: 'PATCH',
    headers: {
      Authorization: `Token token=${store.user.token}`,
    },
    data,
  });
};
```
events for update
```
const onUpdateProfile = function(event) {
  event.preventDefault();

  let data = getFormFields(event.target);

  api.updateProfile(store.user.id, data)
    .then(ui.onUpdateProfile)
    .catch(ui.failure);
};
```

This is my back end update in profile controller
```
  def update
    if @profile.update(profile_params)
      head :no_content
    else
      render json: @profile.errors, status: :unprocessable_entity
    end
  end
```

and just incase this is my patch script on the back end
```
#!/bin/bash

API="${API_ORIGIN:-http://localhost:4741}"
URL_PATH="/profiles/${ID}"
curl "${API}${URL_PATH}" \
  --include \
  --request PATCH\
  --header "Content-Type: application/json" \
  --header "Authorization: Token token=$TOKEN" \
  --data '{
    "profile": {
      "username": "'"${USERNAME}"'",
      "favorite_genre": "'"${FAVORITE_GENRE}"'",
      "favorite_band": "'"${FAVORITE_BAND}"'"
    }
  }'
```

**Comment by bernardlee** [comment]
This error message tells you that the problem is on the back end.

```NoMethodError (undefined method `update' for nil:NilClass):```

Read the error message carefully and think about what the `nil:NilClass` part is referring to and what might have caused this.

**Comment by novakw267** [comment]
I was starring at that in particular as well. I assume I am calling profile incorrectly on the backend, but it worked in node so I wasn't sure what to change that to.

**Comment by bernardlee** [comment]
Do you mean `bin/rails console` when you say `node`?  If so, then testing something in the Rails console using a command like `profile.update(...)` or `User.create(...)` isn't the same thing as sending the back end an actual request.

**Comment by novakw267** [comment]
Fair enough, and no I meant terminal if that's helpful. I also think I noticed something that I want to check when I get home, the ui. Is calling the wrong function.

**Comment by raq929** [comment]
Does the curl request work?
If so, compare the data you're sending via curl with the data you're sending via the front end. They will have to look exactly the same in order for this to work. This likely means making sure that your inputs have the name attribute in the html, because the name attribute is what `getFormFields` uses to create your data object.

**Comment by novakw267** [comment]
So apparently the curl in terminal doesn't work now... Okay so to the back end I think I have something wrong with the class declaration on the back end.

**Comment by novakw267** [comment]
I get two separate errors depending on if I try to patch the profile on the front end or the back end so now I am really unsure where to look? 

front end
```
Completed 500 Internal Server Error in 19ms (ActiveRecord: 7.6ms)


  
NoMethodError (undefined method `update' for nil:NilClass):
  
app/controllers/profiles_controller.rb:33:in `update'
```
back end 
```
Started PATCH "/profiles/" for 127.0.0.1 at 2017-02-24 16:27:50 -0500
  
ActionController::RoutingError (No route matches [PATCH] "/profiles"):
  
actionpack (5.0.1) lib/action_dispatch/middleware/debug_exceptions.rb:53:in `call'
actionpack (5.0.1) lib/action_dispatch/middleware/show_exceptions.rb:31:in `call'
railties (5.0.1) lib/rails/rack/logger.rb:36:in `call_app'
railties (5.0.1) lib/rails/rack/logger.rb:24:in `block in call'
activesupport (5.0.1) lib/active_support/tagged_logging.rb:69:in `block in tagged'
activesupport (5.0.1) lib/active_support/tagged_logging.rb:26:in `tagged'
activesupport (5.0.1) lib/active_support/tagged_logging.rb:69:in `tagged'
railties (5.0.1) lib/rails/rack/logger.rb:24:in `call'
actionpack (5.0.1) lib/action_dispatch/middleware/request_id.rb:24:in `call'
rack (2.0.1) lib/rack/runtime.rb:22:in `call'
activesupport (5.0.1) lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
actionpack (5.0.1) lib/action_dispatch/middleware/executor.rb:12:in `call'
actionpack (5.0.1) lib/action_dispatch/middleware/static.rb:136:in `call'
rack (2.0.1) lib/rack/sendfile.rb:111:in `call'
rack-cors (0.4.0) lib/rack/cors.rb:80:in `call'
railties (5.0.1) lib/rails/engine.rb:522:in `call'
puma (3.7.0) lib/puma/configuration.rb:226:in `call'
puma (3.7.0) lib/puma/server.rb:578:in `handle_request'
puma (3.7.0) lib/puma/server.rb:415:in `process_client'
puma (3.7.0) lib/puma/server.rb:275:in `block in run'
```

**Comment by bernardlee** [comment]
> front end

This one was already discussed earlier in the thread.  https://github.com/ga-wdi-boston/full-stack-project/issues/706#issuecomment-281876320

> back end

Read the stack trace very carefully.  It tells you exactly what the problem is.

**Comment by novakw267** [comment]
Bernard did you mean to link the same issue that we're on? Because I haven't figured it out yet. I literally have no idea I have read it over and over again and tried different iterations of my update in the controller, confirmed my patch script is written correctly and the routes are there. I am just not sure what else to check.

**Comment by bernardlee** [comment]
Yes I did.  What are doing to get those responses that you've pasted?

**Comment by novakw267** [comment]
I will get your some more information, but now I am encountered another error I have to figure out to even get back to that point. I can't even call the command now. 
```
parse_git_branch: command not found
```
So I think I messed something up in my repo.... 

but 
Profiles controller
```# frozen_string_literal: true

# :nodoc:
# Inheriting from open read controller means that a profile should only be
# changable by the user who its attached to.
class ProfilesController < OpenReadController
  before_action :set_profile, only: [:show, :update, :destroy]
  # GET /profiles
  def index
    @profiles = Profile.all

    render json: @profiles
  end

  # GET /profiles/1
  def show
    render json: @profile
  end

  # POST /pfiles
  def create
    @profile = current_user.build_profile(profile_params)

    if @profile.save
      render json: @profile, status: :created
    else
      render json: @profile.errors, status: :unprocessable_entity
    end
  end

  # PATCH/PUT /profiles/1
  # This is for users to be able to change their profile name and other parts
  # their profile.
  def update
    if @profile.update(profile_params)
      head :no_content
    else
      render json: @profile.errors, status: :unprocessable_entity
    end
  end

  # DELETE /profiles/1
  def destroy
    @profile.destroy
  end

  private

  # Use callbacks to share common setup or constraints between actions.
  def set_profile
    @profile = Profile.find_by(id: params[:id], user: current_user)
  end

  # Only allow a trusted parameter "white list" through.
  def profile_params
    params.require(:profile).permit(:username, :favorite_band, :favorite_genre)
  end
end
```
serializer 
```
# frozen_string_literal: true

# :nodoc:
class ProfileSerializer < ActiveModel::Serializer
  attributes :id, :username, :favorite_genre, :favorite_band, :editable
  def editable
    scope == object.user
  end
end
```
Model
```
# frozen_string_literal: true

# :nodoc:
class Profile < ApplicationRecord
  belongs_to :user
  has_many :blog_templates
  validates :username, presence: true
  validates :favorite_genre, presence: true
  validates :favorite_band, presence: true
end
```
routes
```
# frozen_string_literal: true
Rails.application.routes.draw do
  resources :blog_templates, except: [:new, :edit]
  resources :profiles, except: [:new, :edit]
  resources :examples, except: [:new, :edit]
  post '/sign-up' => 'users#signup'
  post '/sign-in' => 'users#signin'
  delete '/sign-out/:id' => 'users#signout'
  patch '/change-password/:id' => 'users#changepw'
  resources :users, only: [:index, :show]
end
``` 
I am going to look into that other error, I think I know why this happening now and I have a 1 on1 with ben today. So hopefully I can figure that other part out. I think it has something to do with the controller, based on the error, the internet, and what you guys have said. As far as I can tell the routes is fine, and the relationship works between user and profile. 

On the terminal I had been basically signing in to get a token, passing the token, and what I want to update to my profile patch script. 

It usually looked like this, though I can't even test to see if that's the correct command I have been using, I can't currently test it from the other error listed above. 
```
TOKEN='insert token' USERNAME='whatever I want to change it to' FAVORITE_GENRE='whatever' FAVORITE_BAND='example' sh './scripts/profiles/update.sh' 
```
Thanks again, I am sorry to be a pain. I am just feeling stumped.

**Comment by bernardlee** [comment]
Run `bin/rails routes` and paste the output.

**Comment by novakw267** [comment]
The back end works, I realized I was not passing the incorrect ID. And I am getting the same error from the front end now I think I am not correctly passing the right ID for the profile attached to the USER. I will run the bin/rails routes and respond in a second too.

```
$ bin/rails routes
        Prefix Verb   URI Pattern                    Controller#Action
blog_templates GET    /blog_templates(.:format)      blog_templates#index
               POST   /blog_templates(.:format)      blog_templates#create
 blog_template GET    /blog_templates/:id(.:format)  blog_templates#show
               PATCH  /blog_templates/:id(.:format)  blog_templates#update
               PUT    /blog_templates/:id(.:format)  blog_templates#update
               DELETE /blog_templates/:id(.:format)  blog_templates#destroy
      profiles GET    /profiles(.:format)            profiles#index
               POST   /profiles(.:format)            profiles#create
       profile GET    /profiles/:id(.:format)        profiles#show
               PATCH  /profiles/:id(.:format)        profiles#update
               PUT    /profiles/:id(.:format)        profiles#update
               DELETE /profiles/:id(.:format)        profiles#destroy
      examples GET    /examples(.:format)            examples#index
               POST   /examples(.:format)            examples#create
       example GET    /examples/:id(.:format)        examples#show
               PATCH  /examples/:id(.:format)        examples#update
               PUT    /examples/:id(.:format)        examples#update
               DELETE /examples/:id(.:format)        examples#destroy
       sign_up POST   /sign-up(.:format)             users#signup
       sign_in POST   /sign-in(.:format)             users#signin
               DELETE /sign-out/:id(.:format)        users#signout
               PATCH  /change-password/:id(.:format) users#changepw
         users GET    /users(.:format)               users#index
          user GET    /users/:id(.:format)           users#show
```

**Comment by bernardlee** [comment]
Not passing in the `id` is why you were getting the error
``` 
ActionController::RoutingError (No route matches [PATCH] "/profiles"):
```
because no such route matches `PATCH /profiles`, but there is a matching route for `PATCH /profiles/:id`.

**Comment by novakw267** [comment]
You're absolutely correct, but I think now I am just passing the incorrect ID for the user, maybe? I know once I figured out what the correct ID for the user on the backend was it worked. I think it maybe similar to what happened to lauren
here: https://github.com/ga-wdi-boston/full-stack-project/issues/633

I am going to go through that a bit again and see if I can  notice anything, unless you have any other suggestions?

I could be way off base here, so I will definitely take any other thoughts.

**Comment by raq929** [comment]
You will need to send the profile ID to patch the profile, not the user id

**Comment by novakw267** [comment]
So sitting down with ben today I am pretty sure it is still passing the user and profile ID but they do not match.  It seems that the backend is looking for them to match. So I am pretty sure I am running into the same issue  that lauren did if the issue I referenced earlier. What was settled on to look into a fix on the front end or the back end?

```
  Parameters: {"profile"=>{"username"=>"ab", "favorite_genre"=>"ab", "favorite_band"=>"ab"}, "submit"=>"Update Profile", "id"=>"31"}
  User Load (0.2ms)  SELECT  "users".* FROM "users" WHERE "users"."token" = $1 LIMIT $2  [["token", "7a928ef6cfcb1555ebe0fb66b441afe5"], ["LIMIT", 1]]
  Profile Load (0.3ms)  SELECT  "profiles".* FROM "profiles" WHERE "profiles"."id" = $1 AND "profiles"."user_id" = 31 LIMIT $2  [["id", 31], ["LIMIT", 1]]
```

**Comment by bernardlee** [comment]
What do your models' associations look like?

**Comment by novakw267** [comment]
user model
```
# frozen_string_literal: true

# :nodoc:
class User < ApplicationRecord
  include Authentication
  has_many :examples
  has_one :profile
end
```

profiles model
```
# frozen_string_literal: true

# :nodoc:
class Profile < ApplicationRecord
  belongs_to :user
  has_many :blog_templates
  validates :username, presence: true
  validates :favorite_genre, presence: true
  validates :favorite_band, presence: true
end
```
blog template model
```
# frozen_string_literal: true

# :nodoc:
class BlogTemplate < ApplicationRecord
  belongs_to :profile
  delegate :blog_templates, to: :profiles
  validates :blog_title, presence: true
  validates :subject, presence: true
  validates :body, presence: true
end
```

**Comment by bernardlee** [comment]
Why does `BlogTemplate` belong to `Profile` instead of `User`?

**Comment by novakw267** [comment]
In my head I did that because I just wanted user to an authorization. So that your profile would determine what you had access to editing. However, maybe I miss understood how that worked? I can change that, but that was my intent.

**Comment by bernardlee** [comment]
Read the code for `OpenReadController` and `ProtectedController` to see what you need.  Make sure that you understand them and set all of your custom controllers to inherit from one of those.

**Comment by novakw267** [comment]
Okay, I get where you coming from and I will definitely look into that again and double check those things, however for the purpose of the issue, (and this is a legit question not me trying to be rude) isn't my relationship between profile and blog-template for another issue? This was originally about not being to update the profile once its made. I haven't even touched trying to crud on the blog template yet. 

That being said is there anything specific in the open read and protected controller you want me to look into.

Side, side note is it possible instead of spanning this long issue, we can just sit down and look this over at some point, since I am getting low on time to figure this out?

**Comment by bernardlee** [comment]
Yes it should be another issue.  Look into everything about those two controllers and how they work.  If one of us is free during workshop time, then we can look into it, but it should be a short meeting.

**Comment by novakw267** [comment]
Okay, ben and I spent time on it yesterday and couldn't figure it out. That is the only reason I asked.

**Comment by novakw267** [comment]
So does a protected controller, only deny crud until authorization happens? And openRead just denies update and destroy unless your authorization is associated with the content?

**Comment by bernardlee** [comment]
Open a new issue for other questions if this one has been resolved.

**Comment by novakw267** [comment]
It has not been resolved.

**Comment by bernardlee** [comment]
The original issue is that you're getting a `NoMethodError` when trying to update a profile, is it not?

**Comment by novakw267** [comment]
It is still the same error I get.

**Comment by bernardlee** [comment]
Was it not resolved when we identified the problem being the `id` not being sent to the server?

**Comment by novakw267** [comment]
So I originally opened the issue for getting that error on the front end. The command works on the back end, I noticed that by passing the wrong id I will get the same error on the back end. 
So assuming actually is the resolution I guess technically its solved, but I do not know how to confirm that on the front end. I had to manually add the correct id in terminal when I ran the command. I am not sure how to do that when the command gets called on the front end. 

So long story short, I have an idea what I need to add to fix it, but I do not know to implement it or if it will work.
